### PR TITLE
Feature/us49/search patient by name

### LINKED ISSUE #49: As a doctor i want to be able to search for a patient by name, (in order later to be redirected to his dossier)

### PR DESCRIPTION
**Pull request by Acetoshi**
Fixes #49 

image

- Re-used existing Search Bar
- Created a new query in Patient resolver
- Created a duplicate test seeder (will need refacto in the future)
- Coded tests for it, with edge cases including basic SQLi and bad inputs
- Created a PatientSearch bar with a dropdown menu made of buttons that call handlePatientSelected(patientId) so it can be used in various situations

### CODE REVIEW

File: client/src/components/SearchBar/SearchBar.tsx [hunk 1]
```diff
@@ -3,7 +3,7 @@ import SearchIcon from '../../icons/SearchIcon';
 
 export default function SearchBar({ handleChange }: SearchBarProps) {
   return (
-    <section className="flex w-full justify-center">
```
Comment (Acetoshi) [inline client/src/components/SearchBar/SearchBar.tsx:6]: section was removed as it makes no semantic sense here

**Review by julien44830 — DISMISSED** [review]

File: client/src/components/PatientSearchBar/PatientSearchBar.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  const [search, setSearch] = useState<string>('');
+  const [getPatientsByname, { data }] = useGetPatientsByNameLazyQuery();
+
+  const handleChange = (value: string): void => {
+    setSearch(value);
+  };
+
+  useEffect(() => {
+    if (search !== '') {
+      getPatientsByname({ variables: { search: search } });
```
Comment (Ektoplasme) [inline client/src/components/PatientSearchBar/PatientSearchBar.tsx]: shouldn't we use a debounce here? so if dr Convergence has trouble writting a patient's name, it doesn't spam the server with queries

File: client/src/main.tsx [hunk 3]
```diff
@@ -44,6 +45,10 @@ const router = createBrowserRouter([
         path: 'rechercher/service/docteur',
         element: <AgentDepartmentDoctor />
       },
+      {
+        path: 'dossiers',
+        element: <DossierBrowser />
```
Comment (Ektoplasme) [inline client/src/main.tsx:50]: Dossier = french Browser = eng, so we have to choose

File: core_api/src/modules/patient/patient.resolver.ts [hunk 4]
```diff
@@ -13,4 +14,19 @@ export default class PatientResolver {
...
     });
   }
+
+  // TODO : rescrtict access to role === doctor | secretary
+  // TODO : limit the number of results ?
+
+  // needed to browse patients by their first or lastname, case insensitive
+  @Query(() => [Patient])
+  async patients(@Arg('search') search: string) {
+    if (search === '' || search === ' ') return [];
```
Comment (Ektoplasme) [inline core_api/src/modules/patient/patient.resolver.ts]: and if search = "  _____"  (lots of whitespaces) ?

File: client/src/components/PatientSearchBar/PatientSearchBar.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+}: PatientSearchBarProps) {
+  const [search, setSearch] = useState<string>('');
+  const [getPatientsByname, { data }] = useGetPatientsByNameLazyQuery();
+
+  const handleChange = (value: string): void => {
+    setSearch(value);
+  };
+
+  useEffect(() => {
+    if (search !== '') {
```
Comment (teasmade) [inline client/src/components/PatientSearchBar/PatientSearchBar.tsx]: No need to not-equals on an empty string? Just if (search) no?

File: core_api/src/modules/patient/patient.resolver.ts (hunk 4 above)
Comment (teasmade) [reply core_api/src/modules/patient/patient.resolver.ts]: Faut "sanitise" le search term avec `.trim()` pour gérer le whitespace: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim

Comment (teasmade) [reply core_api/src/modules/patient/patient.resolver.ts]: ...In fact probably in the front in the useEffect before checking whether to fire the search query...

File: core_api/src/modules/patient/patient.resolver.ts [hunk 6]
```diff
@@ -13,4 +14,19 @@ export default class PatientResolver {
...
+    return await Patient.find({
+      where: [
+        { firstname: ILike(`${search}%`) },
```
Comment (teasmade) [inline core_api/src/modules/patient/patient.resolver.ts:28]: ~~It looks (?) from the TypeORM docs like this will do an AND on the two where conditions, not an OR: https://orkhan.gitbook.io/typeorm/docs/find-options~~

Ignore, I can't read docs closely enough it seems... separate objects give OR, multiple key / values in one object give AND... 👍

**Comment by teasmade** [comment]
Re. refacto for the seed, making it modular and breaking out helper functions is clearly needed, all the more so now we have two different seed scripts. Maybe this can be done alongside the final full dataset seed when that's tackled.

File: client/src/main.tsx (hunk 3 above)
Comment (Acetoshi) [reply client/src/main.tsx:50]: dossier is English as well

File: client/src/components/PatientSearchBar/PatientSearchBar.tsx (hunk 2 above)
Comment (Acetoshi) [reply client/src/components/PatientSearchBar/PatientSearchBar.tsx]: in another PR ?

File: client/src/components/PatientSearchBar/PatientSearchBar.tsx (hunk 5 above)
Comment (Acetoshi) [reply client/src/components/PatientSearchBar/PatientSearchBar.tsx]: for readibility

File: core_api/src/modules/patient/patient.resolver.ts (hunk 4 above)
Comment (Acetoshi) [reply core_api/src/modules/patient/patient.resolver.ts]: Done in frontend

**Review by julien44830 — DISMISSED** [review]

**Review by teasmade — DISMISSED** [review]

**Review by Adri-bis — DISMISSED** [review]

File: client/src/components/PatientSearchBar/PatientSearchBar.tsx [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  const [getPatientsByname, { data }] = useGetPatientsByNameLazyQuery();
+
+  const handleChange = (value: string): void => {
+    setSearch(value);
+  };
+
+  useEffect(() => {
+    const sanitisedSearch = search.trim();
+    if (sanitisedSearch && search === debouncedSearch) {
+      getPatientsByname({ variables: { search: sanitisedSearch } });
```
Comment (Ektoplasme) [inline client/src/components/PatientSearchBar/PatientSearchBar.tsx:21]:
```
  useEffect(() => {
      const sanitisedSearch = debouncedSearch.trim();
      if (sanitisedSearch) {
          getPatientsByname({ variables: { search: sanitisedSearch } });
      }
  }, [debouncedSearch]);

```

**Review by Ektoplasme — APPROVED** [review]

**Review by julien44830 — APPROVED** [review]
